src/libcore/pin.rs

```diff
@@ -306,13 +466,17 @@ impl<'a, T: ?Sized> Pin<&'a mut T> {
     ///
     /// For example, if you  wanted to get a `Pin` of a field of something,
     /// you could use this to get access to that field in one line of code.
+    /// However, there are several gotchas with these "pinning projections";
+    /// see the [`pin` module] documentation for further details on that topic.
     ///
     /// # Safety
     ///
     /// This function is unsafe. You must guarantee that the data you return
     /// will not move so long as the argument value does not move (for example,
     /// because it is one of the fields of that value), and also that you do
     /// not move out of the argument you receive to the interior function.
+    ///
+    /// [`pin` module]: ../../std/pin/index.html#projections-and-structural-pinning
     #[stable(feature = "pin", since = "1.33.0")]
     pub unsafe fn map_unchecked_mut<U, F>(self: Pin<&'a mut T>, func: F) -> Pin<&'a mut U> where
```

Comment:
This function lacks a usage example (we should make sure that there's an example for every function in this module).

Reply:
TBH I never got the point of these usage examples. I don't intend to add them here myself. But if someone wants to do that, feel free to submit a PR against my branch.

Reply:
So that you understand better how to use the function in a larger context (and it also acts as a unit test). For example, in this case, you could illustrate bad usage patterns and good ones. Some people understand textual descriptions better and some people understand examples better (see e.g. https://doc.rust-lang.org/stable/rust-by-example/).

I would file improvements with examples, but I feel there are more qualified people to do so.

Reply:
> So that you understand better how to use the function in a larger context

Hardly any usage example does that IMO, because they just construct some artificial context.

> Some people understand textual descriptions better and some people understand examples better

Fair. But also we don't have to make all imaginable improvements to these docs in one PR. 
In general, these docs PRs tend to be *lots of work*, because when you make any improvement people jump on and quickly suggest tons of other things. That's great, but I have to say stop somewhere or I'll just never get done. The scope creep isn't great. It's much easier to drop some remarks about add this or that, than to actually do it.

Reply:
FWIW, this is a massive step up already, and you're awesome for doing it!

Reply:
> Fair. But also we don't have to make all imaginable improvements to these docs in one PR.

Sure; if you think this is too much and scope-creep, consider these review comments then as post-it notes for someone (possibly myself) to work on later.

> It's much easier to drop some remarks about add this or that, than to actually do it.

Certainly... I'm not dropping these remarks because its easy but because this would be useful for myself as I find pinning generally hard to understand. I was hoping that someone more qualified could write the docs. I.e. it would take much less time for familiar people than for e.g. me.

Reply:
> Hardly any usage example does that IMO, because they just construct some artificial context.

I strongly disagree with this; usage examples are part of what sets Rust documentation apart from docs in many other languages. E.g. [`Iterator::map`](https://doc.rust-lang.org/nightly/std/iter/trait.Iterator.html#method.map) provides useful examples that makes understanding quick at a glance.

Reply:
I would tend to agree that for the purposes of documenting `Pin`, usage examples might not really be all that necessary -- especially currently -- as anyone writing unsafe code interacting with Pin would likely need to internalize extensive documentation already. That makes the usage example less helpful.

I also think it makes sense to hold off until we have more experience in the broader community and as such a better sense of what usage will actually happen in practice.

Reply:
Okay, thanks. That helps clarify things.

FWIW, if you think of me as "more qualified" for this, I really wouldn't know how to write enlightening usage examples. You can't really understand this API function-by-function, I feel -- you need a more holistic view. That's why I focused on the module-level docs.

Having a more extensive case study there would be nice, but I don't think it is realistic for me to get that up and polished until Friday.